Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -2924,26 +2940,277 @@ PhysicalParticleContainer::PushPX (WarpXParIter& pti,
+    }
+
+    // Loop over the particles and update their momentum
+    const amrex::ParticleReal q = this->charge;
+    const amrex::ParticleReal m = this-> mass;
+
+    const auto pusher_algo = WarpX::particle_pusher_algo;
+    const auto do_crr = do_classical_radiation_reaction;
+#ifdef WARPX_QED
+    const auto do_sync = m_do_qed_quantum_sync;
+    amrex::Real t_chi_max = 0.0;
+    if (do_sync) { t_chi_max = m_shr_p_qs_engine->get_minimum_chi_part(); }
+
+    QuantumSynchrotronEvolveOpticalDepth evolve_opt;
+    amrex::ParticleReal* AMREX_RESTRICT p_optical_depth_QSR = nullptr;
+    const bool local_has_quantum_sync = has_quantum_sync();
+    if (local_has_quantum_sync) {
+        evolve_opt = m_shr_p_qs_engine->build_evolve_functor();
+        p_optical_depth_QSR = pti.GetAttribs(particle_comps["opticalDepthQSR"]).dataPtr()  + offset;
+    }
+#endif
+
+    const auto t_do_not_gather = do_not_gather;
+
+    enum exteb_flags : int { no_exteb, has_exteb };
+    enum qed_flags : int { no_qed, has_qed };
+
+    int exteb_runtime_flag = getExternalEB.isNoOp() ? no_exteb : has_exteb;
+#ifdef WARPX_QED
+    int qed_runtime_flag = (local_has_quantum_sync || do_sync) ? has_qed : no_qed;
+#else
+    int qed_runtime_flag = no_qed;
+#endif
+
+    // Using this version of ParallelFor with compile time options
+    // improves performance when qed or external EB are not used by reducing
+    // register pressure.
+    amrex::ParallelFor(TypeList<CompileTimeOptions<no_exteb,has_exteb>,
+                                CompileTimeOptions<no_qed  ,has_qed>>{},
+                       {exteb_runtime_flag, qed_runtime_flag},
+                       np_to_push, [=] AMREX_GPU_DEVICE (long ip, auto exteb_control,
+                                                         auto qed_control)
+    {
+        // Position advance starts from the position at the start of the step
+        // but uses the most recent velocity.
+#if (AMREX_SPACEDIM >= 2)
+        amrex::ParticleReal xp = x_n[ip];
+        amrex::ParticleReal xp_n = x_n[ip];
+#else
+        amrex::ParticleReal xp = 0._rt;
+        amrex::ParticleReal xp_n = 0._rt;
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+        amrex::ParticleReal yp = y_n[ip];
+        amrex::ParticleReal yp_n = y_n[ip];
+#else
+        amrex::ParticleReal yp = 0._rt;
+        amrex::ParticleReal yp_n = 0._rt;
+#endif
+        amrex::ParticleReal zp = z_n[ip];
+        amrex::ParticleReal zp_n = z_n[ip];
+
+        UpdatePositionImplicit(xp, yp, zp, ux_n[ip], uy_n[ip], uz_n[ip], ux[ip], uy[ip], uz[ip], 0.5_rt*dt);
+        setPosition(ip, xp, yp, zp);
+
+        amrex::ParticleReal Exp = 0._rt, Eyp = 0._rt, Ezp = 0._rt;
+        amrex::ParticleReal Bxp = 0._rt, Byp = 0._rt, Bzp = 0._rt;
+
+        if(!t_do_not_gather){
+            // first gather E and B to the particle positions
+            doGatherShapeNImplicit(xp_n, yp_n, zp_n, xp, yp, zp, Exp, Eyp, Ezp, Bxp, Byp, Bzp,
```

Comment:
My understanding is that this function will always gather with Esirkepov-like shape factor, and that this will therefore be energy conserving when using the Esirkepov current deposition. However, when using the direct current deposition (which also has a new implicit version in `CurrentDeposition.H`), I don't think we will get energy conservation.

Did I misunderstand? Or should we had an `if` statement here to use the correct shape function in the case of the direct current deposition for direct deposition? Or do we simply do not want to use the direct deposition with the implicit scheme? (in which case we could maybe revert the corresponding changes in `CurrentDeposition.H`.)

Reply:
Ah, actually now I see that, when using `galerkin_interpolation = 0`, this will use the regular shape factors.